docs/yaml/objects/compiler.yaml

```diff
@@ -586,3 +586,95 @@ methods:
     gcc or msvc, but use the same argument syntax as one of those two compilers
     such as clang or icc, especially when they use different syntax on different
     operating systems.
+
+- name: symbol_visibility_header
+  returns: file
+  since: 0.63.0
+  posargs:
+    name:
+      type: str
+      description: Header file name.
+  kwargs:
+    api:
+      type: str
+      default: '<PROJECT_NAME>_API'
+      description: |
+        Name used for public APIs.
+    compilation:
+      type: str
+      default: '<PROJECT_NAME>_COMPILATION'
+      description: |
+        Name defined only when compiling the library.
+    static_compilation:
+      type: str
+      default: '<PROJECT_NAME>_STATIC_COMPILATION'
+      description: |
+        Name defined only when compiling or using static library.
+    static_only:
+      type: bool
+      default: false
+      description: |
+        If set to true, '<PROJECT_NAME>_STATIC_COMPILATION' is defined inside the
+        generated header.
+  description: |
+    Generate a header file that define a macro to be used to mark all public APIs
+    of a library. Depending on the type of compiler and the platform, this will
+    typically use `__declspec(dllexport)`, `__declspec(dllimport)` or
+    `__attribute__((visibility("default")))`. It is only supported for C and C++
+    compilers.
```

Comment:
Instead of producing different files for different compilers, could we not make it always do the same thing with the appropriate ifdefs? It's not terribly complicated after all. Wit that change this would not need to live inside the compiler object, which is a slightly unintuitive place for it.

Reply:
I said the same thing in my review above. @xclaesse acceded to my suggestion.

The actual implementation in `compiler.py` now writes the same file everywhere, after taking into account the fact that it inserts the string values of the method kwargs into that file.

Effectively, it is now `configure_file()` with a builtin template.

P.S. 

> It's not terribly complicated after all.

This is also why I believe it's best described in a FAQ rather than implemented as a Meson function. My personal feelings on the matter are that users should never use this -- it does pointless file I/O at build time for something that could be checked into git.

But apparently people really want to "have meson guarantee it is correct", for the same reason that when using cmake they want cmake to generate a similar file "to have cmake guarantee it is correct".

Reply:
Yes, everyone gets it wrong, today I discovered that GLib got it wrong forever too. And it seems that even this PR got it wrong according to conversations I had with various people on the subject.

Reply:
I now integrated the changes we made in glib since then, the macro is slightly different.